### ISSUE TITLE
Wrong argument "stream" sent to write_callback()

### ISSUE THREAD

**Issue by segelpaul** [curl/curl#1926]
### I did this

int Counter = 1;
FILE * f = NULL;

    curl_easy_setopt(curl, CURLOPT_URL, MyURL);

    if (verbose)
      curl_easy_setopt(curl, CURLOPT_HEADERFUNCTION, header_function);
    curl_easy_setopt(curl, CURLOPT_HEADERDATA, (void *)&Counter);

    curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, write_function);
    if (verbose)
      f1 = fopen("Loginformular.html", "w");
    curl_easy_setopt(curl, CURLOPT_WRITEDATA, f);
    
    res = curl_easy_perform(curl);

# non-relevant code omitted

### I expected the following

If verbose is set all works fine. If not, in write_function the 4th parameter is not NULL as expected, its &Counter.

My workarround: 
    if (verbose) {
      curl_easy_setopt(curl, CURLOPT_HEADERFUNCTION, header_function);
      curl_easy_setopt(curl, CURLOPT_HEADERDATA, (void *)&Counter);
    }

### curl/libcurl version

curl-7.55.1    build: libcurl-vc-x64-release-static-ipv6-sspi-winssl

[curl -V output]

### operating system
Win7   x64

**Comment by bagder** [comment]
However confusing this looks, it is actually intended - and documented. Maybe we need to clarify it better.

If you set `CURLOPT_HEADERDATA` without setting `CURLOPT_HEADERFUNCTION`, the result is that libcurl will instead call the `CURLOPT_WRITEFUNCTION` with the headerdata pointer for each received header.

While as if you also set the `CURLOPT_HEADERFUNCTION`, that callback will be used instead for incoming headers.

**Comment by segelpaul** [comment]
Hi Daniel,

In my case, I set CURLOPT_WRITEDATA and than, I get in write_callback() an data argument that was set with CURLOPT_HEADERDATA before! CURLOPT_WRITEDATA was set.
That is wrong!
When I set CURLOPT_HEADERDATA, it shold apply to header_callback() and CURLOPT_WRITEDATA to write_callback().
Applying CURLOPT_HEADERDATA to write_callback() is very unexpected.

My my application, I was interested in the cookie from a login page only. It's header and content for debug only.




--------------------------------------------
Daniel Stenberg <notifications@github.com> schrieb am Do, 28.9.2017:

 Betreff: Re: [curl/curl] Wrong argument "stream" sent to write_callback() (#1926)
 An: "curl/curl" <curl@noreply.github.com>
 CC: "segelpaul" <reche_j@yahoo.de>, "Author" <author@noreply.github.com>
 Datum: Donnerstag, 28. September, 2017 00:57 Uhr
 
 However confusing this looks, it
 is actually intended - and documented. Maybe we need to
 clarify it better.
 If you set CURLOPT_HEADERDATA without
 setting CURLOPT_HEADERFUNCTION, the result is
 that libcurl will instead call the
 CURLOPT_WRITEFUNCTION with the headerdata
 pointer for each received header.
 While as if you also set the
 CURLOPT_HEADERFUNCTION, that callback will be
 used instead for incoming headers.
 
 —
 You are receiving this because you authored the thread.
 Reply to this email directly, view
 it on GitHub, or mute
 the thread.

**Comment by bagder** [comment]
> In my case, I set CURLOPT_WRITEDATA

Yes, but you *also* set CURLOPT_HEADERDATA. Then your write callback will be called for both headers and body, and for the headers it will get the CURLOPT_HEADERDATA pointer while for the body it will get the CURLOPT_WRITEDATA pointer.

> Applying CURLOPT_HEADERDATA to write_callback() is very unexpected

I  agree that it is quirky, but it is documented and since it is behavior we shipped eons ago we can't easily remove it or risk breaking a few applications out there. If I would do it all over from the start today, I would not have done it this way.

A very simple fix for you is to always set CURLOPT_HEADERDATA and CURLOPT_HEADERFUNCTION at the same time.

**Comment by segelpaul** [comment]
To find out the cause of crash was a long way, it's very unexpected.
The solution is easy IF YOU KNOW this behave. Only CURLOPT_HEADERDATA and CURLOPT_HEADERFUNCTION together or not at all.




--------------------------------------------
Daniel Stenberg <notifications@github.com> schrieb am Do, 28.9.2017:

 Betreff: Re: [curl/curl] Wrong argument "stream" sent to write_callback() (#1926)
 An: "curl/curl" <curl@noreply.github.com>
 CC: "segelpaul" <reche_j@yahoo.de>, "Author" <author@noreply.github.com>
 Datum: Donnerstag, 28. September, 2017 08:53 Uhr
 
 
 In my case, I set CURLOPT_WRITEDATA
 
 Yes, but you also set CURLOPT_HEADERDATA. Then
 your write callback will be called for both headers and
 body, and for the headers it will get the CURLOPT_HEADERDATA
 pointer while for the body it will get the CURLOPT_WRITEDATA
 pointer.
 
 Applying CURLOPT_HEADERDATA to write_callback() is very
 unexpected
 
 I  agree that it is quirky, but it is documented and
 since it is behavior we shipped eons ago we can't easily
 remove it or risk breaking a few applications out there. If
 I would do it all over from the start today, I would not
 have done it this way.
 A very simple fix for you is to always set
 CURLOPT_HEADERDATA and CURLOPT_HEADERFUNCTION at the same
 time.
 
 —
 You are receiving this because you authored the thread.
 Reply to this email directly, view
 it on GitHub, or mute
 the thread.
